docs/compartmentalizing_dav1d.md

```diff
@@ -0,0 +1,577 @@
+#include <ia2_compartment_init.inc>
+```
+
+For the other compartment (`src/lib.c`),
+it gets this IA2 declaration:
+
+```c
+#include <ia2.h>
+#define IA2_COMPARTMENT 2
+#include <ia2_compartment_init.inc>
+```
+
+## Rewriting
+
+Next, we must run `ia2-rewriter` on `dav1d` to get the rewritten sources.
+This is tricky, as `dav1d` is built with `meson`,
+while the existing compartmentalized builds in the IA2 repo use `cmake`.
+There are a lot of undocumented steps in that,
+and things that `meson` does differently, so they're documented here:
+
+### Generating a `compile_commands.json`
+
+`ia2-rewriter` is a `libclangTooling` tool,
+so it operates on a `compile_commands.json` compilation database.
+`meson`, like `cmake`, can generate a `compile_commands.json`,
+and does so by default when running `meson setup`.
+
+#### Absolute Paths and `canonicalize_compile_command_paths.py`
+
+However, it generates a `compile_commands.json` with relative paths.
+`libclangTooling` hates relative paths and usually does the incorrect thing with them.
+To get around this, we added a simple script,
+[`canonicalize_compile_command_paths.py`](../tools/rewriter/canonicalize_compile_command_paths.py),
+which rewrites a `compile_commands.json`'s paths to be absolute (and canonical/fully resolved).
+This includes paths that can be detected within arguments,
+though this is detected as a best effort.
+Because `libclangTooling` requires a compilation database
+to be named exactly `compile_commands.json`,
+`canonicalize_compile_command_paths.py` assumes this and operates on
+a `compile_commands.json` in the current working directory.
+
+#### Other Flags
+
+`ia2-rewriter` also requires some other flags to be set.
+Some of these can be set by `libclangTooling`'s `--extra-arg` argument,
+those apply to all of the compile commands at once,
+so when we want different arguments to be added to different compile commands,
+we need to add these to the build system, to `meson`,
+in order to have them show up in the `compile_commands.json`.
+
+The main argument that is like this is `-DPKEY`.
+Unlike `#define IA2_COMPARTMENT` and `INIT_RUNTIME`,
+`PKEY` must be defined for every translation unit.
+This is easiest to add in the build system,
+so that when we run `meson setup`, the `compile_commands.json` generated
+contains the correct `-DPKEY`s that `ia2-rewriter` will see.
+
+There are also other args that we need that are the same for all compile commands.
+So these could be specified with `--extra-arg`,
+but as we're already modifying the `meson.build`s, it's simpler to just modify them there.
+
+First, there are required args:
+
+* `-DIA2_ENABLE=1`
+* `'-I' + join_paths(ia2_path, 'runtime/libia2/include')`
+
+Then there are args for debugging:
+
+* `-DIA2_DEBUG=1` (for debug assertions)
+* `-DIA2_DEBUG_LOG=1` (for verbose logging)
+
+Then there are overrides:
+
+First just disabling some warnings/errors that `dav1d` enabled but IA2's runtime doesn't follow
```

Comment:
I feel like we should support building with these flags at some point.

Reply:
Okay, though `-Wno-unknown-warning-option` we probably want to keep since I think it's a `gcc` vs `clang` thing.